File: 0496-next-greater-element-i/0496-next-greater-element-i.java
import java.util.ArrayList;
class Solution {
    public void print(int[] nums){
        for(int i:nums){
        System.out.print(i+" ");
       }
        System.out.println();
    }
    public void printAl(ArrayList<Integer> nums){
        for(int i:nums){
        System.out.print(i+" ");
       }
        System.out.println();
    }
    
    public void convert(int num[],ArrayList<Integer> n){
        for(int i:num){
            n.add(i);
        }
    }
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        print(nums1);
        print(nums2);
        ArrayList<Integer> n1=new ArrayList<>();
        ArrayList<Integer> n2=new ArrayList<>();
        convert(nums1,n1);
        convert(nums2,n2);
        printAl(n1);
        printAl(n2);
        int k=0;
        int[] num=new int[n1.size()];
        for(int p=0;p<n1.size();p++){
            num[p]=-1;
        }
        for(int i:n1){
            int index=n2.indexOf(i);
            for(int j=index;j<n2.size();j++){
                if (n2.get(j)>i){
                    num[k]=n2.get(j);
                    break;
                }   
            }
            k++;
        }
        return num;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
#idea is montonic stack
#mostly we do it by index but since here everything is unique qe can use hash map for easy mapping
#idea is that once we see a larger element than stack we mean that we have seen a greater el than prev,
#thus we assign next greater el of prev as curr
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        hash = dict()
        for num in nums2:
            while stack and num>stack[-1]:
                prev = stack.pop()
                hash[prev] = num
            stack.append(num)
        for el in stack:
            hash[el] = -1
        return [hash[i] for i in nums1]